From 6553400cc4356ca6e49bb970fcc20dc76eab4ecb Mon Sep 17 00:00:00 2001
From: Bastien Riviere <me@babariviere.com>
Date: Sat, 7 May 2022 12:36:54 +0200
Subject: [PATCH 2/2] debug

---
 modules/shepherd/service.scm | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/modules/shepherd/service.scm b/modules/shepherd/service.scm
index f343938..7d0bc1b 100644
--- a/modules/shepherd/service.scm
+++ b/modules/shepherd/service.scm
@@ -1494,13 +1494,15 @@ permissions for its parent directory."
 (define (wait-for-readable ports)
   "Suspend the current task until one of @var{ports} is available for
 reading."
-  (suspend-current-task
-   (lambda (sched k)
-     (for-each (lambda (port)
-                 (schedule-task-when-fd-readable sched
-                                                 (port-read-wait-fd port)
-                                                 k))
-               ports))))
+  #t
+  ;; (for-each (lambda (port)
+  ;;             (suspend-current-task
+  ;;              (lambda (sched k)
+  ;;                (schedule-task-when-fd-readable sched
+  ;;                                                (port-read-wait-fd port)
+  ;;                                                k))))
+  ;;           ports)
+  )
 
 (define* (make-systemd-constructor command endpoints
                                    #:key
-- 
me@babariviere.com

